### PR TITLE
chore(spa refactor): refactoring dashboard to use api's instead of bootstrapdata

### PR DESCRIPTION
**Pull request by pkdotson**
### SUMMARY
This PR will begins the first steps of moving dashboard into spa while using the new dashboard api's `dashboard/:id/charts` and `/dashboards/:id`.

Thanks for @suddjian for the pairing! :)
### TEST PLAN
updates to existing tests and end tests as need.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/cypress-base/cypress/integration/dashboard/load.test.ts
Comment (rusackas) [inline superset-frontend/cypress-base/cypress/integration/dashboard/load.test.ts:26]: Should this be a `beforeAll` so we're not doing unnecessary repeat logins?

Comment (suddjian) [reply superset-frontend/cypress-base/cypress/integration/dashboard/load.test.ts:26]: I think I tried that and it broke 🤷‍♂️

File: superset-frontend/cypress-base/cypress/integration/dashboard/load.test.ts
Comment (rusackas) [inline superset-frontend/cypress-base/cypress/integration/dashboard/load.test.ts:36]: Would there be any value in testing the other permutations of these params?

Comment (suddjian) [reply superset-frontend/cypress-base/cypress/integration/dashboard/load.test.ts:36]: Sure, I was basically just going for parity with the existing tests that I deleted but I don't see why we can't add more.

File: superset-frontend/cypress-base/cypress/integration/dashboard/nativeFilters.test.ts
Comment (rusackas) [inline superset-frontend/cypress-base/cypress/integration/dashboard/nativeFilters.test.ts:69]: Not a problem, but I'm curious why we need the `.ant-modal` selector here. Mainly wondering if our data-test attribute is too general and should just be made more specific.

Comment (suddjian) [reply superset-frontend/cypress-base/cypress/integration/dashboard/nativeFilters.test.ts:69]: Oh, I think that's a copy-paste artifact

File: superset-frontend/src/chart/chartReducer.ts
Comment (rusackas) [inline superset-frontend/src/chart/chartReducer.ts:25]:
```suggestion
import { HYDRATE_DASHBOARD } from 'src/dashboard/actions/hydrate';
```

File: superset-frontend/src/common/hooks/apiResources/dashboards.ts
Comment (rusackas) [inline superset-frontend/src/common/hooks/apiResources/dashboards.ts:23]: cc @eschutho just to raise awareness of pattern use

File: superset-frontend/src/dashboard/App.jsx
Comment (rusackas) [inline superset-frontend/src/dashboard/App.jsx:40]: Maybe we need a `ProviderProvider` ;)

Comment (suddjian) [reply superset-frontend/src/dashboard/App.jsx:40]: Oh yeah! This should be addressed in the app merge PR.

File: superset-frontend/src/dashboard/actions/hydrate.js
Comment (rusackas) [inline superset-frontend/src/dashboard/actions/hydrate.js:66]: I'm a little fuzzy on what else would be in this array and what we're losing by only extracting the first value.

Comment (suddjian) [reply superset-frontend/src/dashboard/actions/hydrate.js:66]: Multiple url parameters can share the same key (e.g. http://example.com/page?foo=bar&foo=baz). In this case, the values will come through from the `query-string` library as an array instead of a string (an inconsistent api surface). This function is a frontend reimplementation of the `urlParams` processing that [was happening on the backend](https://github.com/apache/superset/pull/13306/files/52b31118b074d3e2bf9c33e1880788c9da985b31#diff-76c6a0b5c14a297c62ac3207416870471affdb153c5ae12b96d7b7da9304e9ddL1869), and throws away all but the first instance of a given url param key, which matches Flask behavior when calling `request.args.items()`.

I'll add a comment to the code.

File: superset-frontend/src/dashboard/actions/hydrate.js
Comment (rusackas) [inline superset-frontend/src/dashboard/actions/hydrate.js:79]: Could/should `useDashboard` do the JSON parsing, so we just get for-real objects from the hook to work with?

Comment (suddjian) [reply superset-frontend/src/dashboard/actions/hydrate.js:81]: good idea!

Comment (rusackas) [reply superset-frontend/src/dashboard/actions/hydrate.js:81]: There's actually a lot of JSON parsing going on in here... if any of it can be moved closer to where the data is initially captured, it might be helpful to reduce the number of consumers that may need to parse it ¯\_(ツ)_/¯

File: superset-frontend/src/dashboard/actions/hydrate.js
Comment (rusackas) [inline superset-frontend/src/dashboard/actions/hydrate.js:118]: var name nit: once parsed, it's just the positionObj or positionValue or something, right?

File: superset-frontend/src/dashboard/util/getPermissions.tsx
Comment (rusackas) [inline superset-frontend/src/dashboard/util/getPermissions.tsx:36]:
```suggestion
  return roleList.filter(([role,perms]) => {
      return perms.filter(item => {
      	return item[0] === perm && item[1] === view
      }).length > 0;
  }).length > 0; 
```
I'm not sure if that's any easier to read, but I had to try...

Comment (suddjian) [reply superset-frontend/src/dashboard/util/getPermissions.tsx:36]: @pkdotson wanna check this out?

Comment (suddjian) [reply superset-frontend/src/dashboard/util/getPermissions.tsx:36]: Maybe it should be memoized too or stored in a map for faster lookup

Comment (pkdotson) [reply superset-frontend/src/dashboard/util/getPermissions.tsx:36]: Yep will check out!

**Comment by ktmud** [comment]
Ephemeral env Preview blocked by Chrome malware detection?

superset-malware

**Comment by robdiciuccio** [comment]
@ktmud Strange. Test env works for me in Chrome versions 89.0.4389.90 and 89.0.4389.114 (latest). Tested in incognito as well. Do you have any 3rd-party malware extensions running? Maybe that IP was previously flagged by Google?

File: superset-frontend/src/components/SupersetResourceSelect/index.tsx
Comment (eschutho) [inline superset-frontend/src/components/SupersetResourceSelect/index.tsx:108]: intentional console.log?

File: superset-frontend/src/common/hooks/apiResources/dashboards.ts
Comment (eschutho) [reply superset-frontend/src/common/hooks/apiResources/dashboards.ts]: Thanks. I guess we'll continue to use component state for now instead of redux until we come to a final decision on the pattern? Obviously we don't want to hold up this SPA effort, but it may mean more refactoring later on.

File: superset-frontend/src/components/SupersetResourceSelect/index.tsx
Comment (suddjian) [reply superset-frontend/src/components/SupersetResourceSelect/index.tsx:108]: oops, you found some debug code!

File: superset-frontend/src/dashboard/actions/hydrate.js
Comment (suddjian) [reply superset-frontend/src/dashboard/actions/hydrate.js]: I moved the metadata and the positions parsing, I think those are the main ones.

**Comment by ktmud** [comment]
@robdiciuccio I'm using Chrome 89.0.4389.114 , too.  Maybe it's just this one IP: http://54.212.52.195:8080 had some bad history.

File: superset-frontend/src/common/hooks/apiResources/dashboards.ts
Comment (suddjian) [reply superset-frontend/src/common/hooks/apiResources/dashboards.ts]: I think we should also start to look at third party libs that have a very similar API to the `useResource` hooks. @ktmud linked some good ones in the original PR: #13218

File: superset-frontend/src/common/hooks/apiResources/dashboards.ts
Comment (rusackas) [inline superset-frontend/src/common/hooks/apiResources/dashboards.ts:21]: If our ol' friend Codecov is to be believed, it looks like both these imports could use some unit tests.

**Comment by junlincc** [comment]
Manually tested all functionalities on Dashboard (view and edit modes)
Everything works expectedly except 
1) Toggle fullscreen on Dashboard, but I can confirmed that it's not related to this PR. 
2) "GLOBAL_ASYNC_QUERIES": True, is not compatible with "DASHBOARD_NATIVE_FILTERS": True, But i believe this was fixed in master 
@robdiciuccio @suddjian 

CC: @rosemarie-chiu
https://user-images.githubusercontent.com/67837651/113641856-47617580-9633-11eb-9c19-796b88158d0e.mov

When both flags are on. 
Screen Shot 2021-04-05 at 5 24 33 PM

**Comment by suddjian** [comment]
I want to make sure this supports the `DASHBOARD_CACHE` feature flag. @ktmud do you think that supporting the feature flag would be as simple as adding a `@cache_manager.cache.memoize` decorator to the new DAO accessors used in the new endpoints, similar to the one on `dashboard.full_data`? Or is there more to it?

**Comment by ktmud** [comment]
> I want to make sure this supports the `DASHBOARD_CACHE` feature flag. @ktmud do you think that supporting the feature flag would be as simple as adding a `@cache_manager.cache.memoize` decorator to the new DAO accessors used in the new endpoints, similar to the one on `dashboard.full_data`? Or is there more to it?

I'd imagine it should work. Just make sure the correct cache key is set and used in the `clear_cache` functions.

**Comment by junlincc** [comment]
> 2\. "GLOBAL_ASYNC_QUERIES": True, is not compatible with "DASHBOARD_NATIVE_FILTERS": True, But i believe this was fixed in master

scratch that. master is also broken when both FF = true

**Comment by suddjian** [comment]
@ktmud Actually I did some testing and it seems like the memoization isn't currently working. I turned the FF on, went to a dashboard, and reloaded the page, and `full_data` got called a second time. I got the same problem when I added caching to the DAO functions. Could you confirm, and would you mind if I skip cache support for now, and we can revisit?

File: superset-frontend/src/dashboard/App.jsx
Comment (ktmud) [inline superset-frontend/src/dashboard/App.jsx:28]:
```suggestion
import DashboardPage from './components/DashboardPage';
```

Let's have more consistent component name? `DashboardPage.tsx` probably also don't need to live under `./components`.

**Comment by ktmud** [comment]
Oh, didn't notice you were reading datasets, charts, and dashboard metadata separately from three different API endpoints... Thought we'd still use `full_data` to bootstrap the Dashboard page. Then it would be a little tricky to persist a backend cache. I think the most expensive part of dashboard bootstrap data is checking dataset permissions, because it needs to run a separate SQL query to get the dataset info for each chart in the dashboard. If we can somehow make this process faster, then maybe a `full_data` cache is not necessary.

Feel free to skip cache support for now and revisit later.

**Review by ktmud — COMMENTED** [review]
Didn't have time to do a full-pass review but noticed one small issue...

File: superset-frontend/src/dashboard/App.jsx
Comment (suddjian) [reply superset-frontend/src/dashboard/App.jsx:28]: Oh good catch, that's an auto-refactor artifact. I will also move `DashboardPage` to /containers.

File: superset-frontend/src/common/hooks/apiResources/dashboards.ts
Comment (rusackas) [reply superset-frontend/src/common/hooks/apiResources/dashboards.ts:21]: Huh... it seems to have settled down... ¯\\\_(ツ)_/¯

**Review by rusackas — APPROVED** [review]
Ok... went through all the code. All of it LGTM, though Python ain't my strong suit, so others may want to weigh in more there. I think the caching that @ktmud mentions will need a revisit, but everything here looks good to go to these eyes. Tested on the ephemeral environment, and I've failed to break it! Thanks for all the hard work here!

**Comment by graceguo-supercat** [comment]
Unfortunately this PR caused dashboard performance issue for airbnb:
Screen Shot 2021-04-20 at 11 03 43 AM

Please see this chart, which is daily median value of all dashboard metadata load time (in airbnb). After metadata is loaded, dashboard will fetch charts and render charts. You will see that after 4/15 (release date) the dashboard metadata load time is increased a lot, up to now the load time increased about ~30%. 

In airbnb we have a lot of big dashboard, which have >100 charts from multiple datasets. The original solution, load full dashboard data in one API and **cache** the dashboard request, was one of proven way which improved dashboard load performance. But this good practice seems will not be carried over.

Given the impact to our performance metric, airbnb is reverted this PR from our own production environment. I hope open source community and Preset team could re-consider this solution. By now we would suggest:
- add feature flag for this SPA solution, so that airbnb can escape from this change
- add cache for all the new API, to reduce the server-side dashboards/datasets/charts request processing time.

@ktmud @pkdotson @junlincc @zuzana-vej

**Comment by junlincc** [comment]
Thanks for reporting, Grace. Could you confirm the perf change is caused by this PR solely? SPA solution is designed to improve the application performance overall and we are happy to optimize the solution so that it can work for your use cases with large amount of data. I'm not sure adding feature flag to it is the best way to go, as we have way too many feature flags nowadays which increase the difficult in testing. 

We requested a sample dataset from Airbnb multiple times, so that we can properly test each PR to make sure they work as expected when the dataset is extremely large.. but we haven't heard back in that sense...........

We acknowledge the potential inconvenience we might have caused, let's work together towards solution. :)

**Comment by suddjian** [comment]
I appreciate the inconvenience of the performance hit, but I don't think a revert is the appropriate response in this case, for the following reasons.

- The plan to skip cache support for now and follow up later was agreed upon by a member of the Airbnb team (as documented above)
- A 25% performance hit, while inconvenient, is not nearly as critical as, for example, a broken page or security issue
- The community was not made aware of this issue before the revert occurred

I'm happy to work with folks on moving forward with adding performance improvements such as a cache, but reverting this change makes the work of developers significantly harder as we have to deal with substantially more source control branching complexities. Considering the circumstances I think a conversation should have been at least attempted.

Adding a feature flag to this project would be impractical, as various details need to change about how the app is loaded and the constant FF checks needed would make issues more likely to occur. Optimizing the endpoints seems more practical.

@ktmud do you have any availability to work together on a caching solution for this problem?

**Comment by graceguo-supercat** [comment]
@junlincc and @suddjian Thanks for the replies. Revert right now only happens in airbnb production, and I do not mean to revert in open source code base, because i agree all points that Aaron mentioned, we also prefer to **_fix forward_**. 

Revert is **_short term solution_** for airbnb. I will keep posting our perf data once the new results ready. Since revert was happened 4/19, it will take us a couple of days to aggregate data (airbnb aggregate superset perf data by _daily_).

For proposed solutions, I also prefer **_cache all API_** than adding feature flag. If there are any progress in this direction, i am happy to bring back this PR and test it again in airbnb.

**Comment by ktmud** [comment]
@suddjian I don't have time to actually work on this but happy to review PRs if needed.

I think one simple fix is to still use `full_data` for bootstrap, and only call datasets and charts API when navigate. Caching at the API level may also be feasible, but how to clear cache properly would be a little tricky.

**Comment by graceguo-supercat** [comment]
Screen Shot 2021-04-20 at 7 51 51 PM

Hey @junlincc and @suddjian We have 1 more data point: once we reverted this PR (this PR and #18184 are the only change), the perf goes back to normal. We didn't run full 24 hours since the revert was happened after UTC 0:00. Can this chart enough to show the perf impact?

As you see, more recent PRs are based on this change, it will be very hard for us to sync weekly with master branch and keep this PR out. Could you help to make some simple fix so that we can move forward, like @ktmud suggested above? Thanks!

**Comment by suddjian** [comment]
Thanks @graceguo-supercat, the perf impact is understood. Fixing this is my current focus. I am going to try to add caching on the endpoint since I think that'd be cleaner, and if it works then I think `full_data` can be removed as nothing else uses it.

**Comment by graceguo-supercat** [comment]
Thanks @suddjian In the new SPA solution, there are 3 extra api calls (to fetch dash/charts/datasets details) that issued **_after_** initial page load, will this extra calls make the overall initial metadata loading + processing faster? this is my biggest concern. IF this is the major blocker, what else we can do?

**Comment by suddjian** [comment]
I actually think loading data after initial page load will enable significant performance wins, when the SPA project is complete. A single page app will not need to load the page at all, so comparatively making three API calls to fetch data for the page will be small. Plus, no blank screen will mean better perceived performance as well.

**Pull request merged**
